float32 x_distance
float32 y_distance
float32 distance